### DISCUSSION TITLE
Dispose

### DISCUSSION THREAD

**Discussion by unicomp23** [esfx/esfx#78] [Q&A]
@rbuckton is dispose supposed to work the same as c# ?  I'm trying to figure out how to use it.

**Comment by rbuckton** [comment]
Essentially, yes. `Disposable.dispose` (which is an alias to [`Symbol.dispose`](https://github.com/tc39/proposal-explicit-resource-management), if it exists) is a symbol-named method that acts as the JS equivalent of the C# `Disposable.Dispose()` method.

In the [Explicit Resource Management proposal](https://github.com/tc39/proposal-explicit-resource-management), you would use a `using x = y` declaration to declare the variable `x`, assign to it the value of `y`, and register `x`'s `Symbol.dispose` method to be invoked when the current block scope exits.

The `@esfx/disposable` package contains an implementation of the `DisposableStack` object from that proposal, which can be used to approximate that behavior via `try...finally`:

```js
import { DisposableStack } from "@esfx/disposable";
const stack = new DisposableStack();
try {
  const x = stack.use(y);
}
finally {
  stack.dispose();
}
```

You can also emulate the _actual_ semantics of `using` via `for..of` and the `Disposable.scope()` helper from `@esfx/disposable`, though it's a bit more work to set up:

```js
//// proposed 'using' syntax:
// {
//   using x = y;
//   ...
// }

// 'Disposable.scope()' translation:
for (const { using, fail } of Disposable.scope()) {
  try {
    const x = using(y);
    ...
  }
  catch (e) {
    fail(e);
  }
}
```

Note that `@esfx/disposable` hasn't yet been updated to the current Stage 3 proposal, which throws a `SuppressedError` rather than an `AggregateError` when there are multiple exceptions during cleanup.

`@esfx/disposable` also implements the [Async Explicit Resource Management proposal](https://github.com/tc39/proposal-async-explicit-resource-management)'s `AsyncDisposableStack`, which is currently at Stage 2.

**Comment by unicomp21** [comment]
Got it, thanks!

**Reply by unicomp23** [reply]
i guess the current async dispose implementation is pretty stable?  or will it change due to stage 4?

**Reply by rbuckton** [reply]
I think it is fairly stable. Generally the only changes between Stage 3 and 4 are due to implementation feedback, or if there is a name change that needs to occur as a result of a web incompatibility.
